### PR TITLE
curing the vulnerable parser

### PR DESCRIPTION
**Pull request by daxtens**
Signed-off-by: Daniel Axtens <dja@axtens.net>

### CODE REVIEW

File: content/blog/2019-04-02-parsers.markdown [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+You can tease out 2 contentions in the first part of the article:
+
+ - there should be a formal langague that describes the data, and
+ - this language should be as simple as possible, ideally being regular and context-free.
+
+Neither of these are bad ideas - in fact they're both good ideas - but I don't know that I draw the same links between them and security.
+
+Consider PostScript as a possible counter-example. It's a Turing-complete language, so it absolutely cannot have predicable results. It has a well documented specification and executes in a restricted virtual machine. So let's say that it satisfies only the first plank of their argument.
+
+I'd say that PostScript has a good security record, despite being Turing complete. PostScript has been around since 1985 and apart from the recent bugs in GhostScript, it doesn't have a long history bugs and exploits. Maybe this just because no-one has really looked, or maybe it is possible to have reasonably safe complex languages by restricting the execution environment, as PostScript consciously and deliberately does.
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: "long history of"

File: content/blog/2019-04-02-parsers.markdown [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+### Is this a useful way to deal with the world as we find it?
+
+My main problem with the general LangSec approach that this article takes is this: to get to their desired world, we need to rewrite a bunch of things with entirely different language foundations. The article talks about HTML and PDFs as examples of unsafe formats, but I cannot imagine the sudden wholesale replacement of either of these - although I would love to be proven wrong.
+
+Can we get even part of the way with existing standards? Kinda-sorta, but mostly no, and to the authors' credit, they are open about this. They argue that formal definition parsing the language should be the "most restrictive input definition" - they specifically require you to "give up attempting to accept arbitrarily complex data", and call for "subsetting of many protocols, formats, encodings and command languages, including eliminating unneeded variability and introducing determinism and static values".
+
+No doubt we would be in a better place if people took up these idea for future programs.
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: ideas

File: content/blog/2019-04-02-parsers.markdown [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+However, for our current set of programs and use cases, this is probably not tractable in any meaningful way.
+
+## The rest of the paper
+The rest of the paper is reasonably interesting. Their general theory is that you should build your parsers based of a formal definition of a language, and that the parser should convert the input data to a set of objects, and then your business logic should deal with those objects. This is the 'recognizer pattern', and is illustrated below:
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: "based off"

Comment (daxtens) [reply content/blog/2019-04-02-parsers.markdown]: went with "based on"

File: content/blog/2019-04-02-parsers.markdown [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+A big theme in this article is predictability:
+
+> Trustworthy input is input with predictable effects. The goal of input-checking is being able to predict the input’s effects on the rest of your program.
+
+This seems sensible enough at first, but leads to some questionable assertions, such as:
+
+> Safety is predictability. When it's impossible to predict what the effects of the input will be (however valid), there is no safety.
+
+They follow this with an example of Ethereum contracts stealing money from the DAO. The example is compelling enough, but again comes with a very strong assertion about the impossibility of securing a languague virtual machine:
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: language

File: content/blog/2019-04-02-parsers.markdown [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+> From the viewpoint of language-theoretic security, a catastrophic exploit in Ethereum was only a matter of time: one can only find out what such programs do by running them. By then it is too late.
+
+I'm not sure that (a) I buy the assertions, or that (b) they provide a useful way to deal with the world as we find it.
+
+### Is this even correct? 
+You can tease out 2 contentions in the first part of the article:
+
+ - there should be a formal langague that describes the data, and
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: language

File: content/blog/2019-04-02-parsers.markdown [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ - this language should be as simple as possible, ideally being regular and context-free.
+
+Neither of these are bad ideas - in fact they're both good ideas - but I don't know that I draw the same links between them and security.
+
+Consider PostScript as a possible counter-example. It's a Turing-complete language, so it absolutely cannot have predicable results. It has a well documented specification and executes in a restricted virtual machine. So let's say that it satisfies only the first plank of their argument.
```
Comment (ajdlinux) [inline content/blog/2019-04-02-parsers.markdown]: predictable

**Pull request merged**